xmtp_id/src/credential_verifier.rs

```diff
@@ -0,0 +1,99 @@
+use prost::Message;
+use xmtp_mls::{
+    credential::{AssociationError, Credential},
+    types::Address,
+};
+use xmtp_proto::xmtp::mls::message_contents::MlsCredential as CredentialProto;
+
+#[derive(Debug, Clone)]
+pub enum AssociationType {
+    ExternallyOwned,
+    #[allow(dead_code)]
+    SmartContract,
+    Legacy,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum VerificationError {
+    #[error(
+        "Address mismatch in Association: Provided:{provided_addr:?} != signed:{signing_addr:?}"
+    )]
+    AddressMismatch {
+        provided_addr: Address,
+        signing_addr: Address,
+    },
+    #[error("Installation public key mismatch")]
+    InstallationPublicKeyMismatch,
+    #[error("protobuf deserialization: {0}")]
+    Deserialization(#[from] prost::DecodeError),
+    #[error("bad association: {0}")]
+    BadAssocation(#[from] AssociationError),
+}
+
+pub struct VerifiedCredential {
+    pub account_address: String,
+    pub account_type: AssociationType,
+}
+
+impl VerifiedCredential {
+    pub fn account_address(&self) -> &String {
+        &self.account_address
+    }
+}
+
+pub struct VerificationRequest {
+    installation_public_key: Vec<u8>,
+    credential: Vec<u8>,
+}
+
+impl VerificationRequest {
+    pub fn new<I, C>(installation_public_key: I, credential: C) -> Self
+    where
+        I: AsRef<[u8]>,
+        C: AsRef<[u8]>,
+    {
+        Self {
+            installation_public_key: installation_public_key.as_ref().to_vec(),
+            credential: credential.as_ref().to_vec(),
+        }
+    }
+}
+
+type VerificationResult = Result<VerifiedCredential, VerificationError>;
+
+#[async_trait::async_trait]
+pub trait CredentialVerifier {
+    async fn verify_credential(request: VerificationRequest) -> VerificationResult;
+    async fn batch_verify_credentials(
+        credentials_to_verify: Vec<VerificationRequest>,
+    ) -> Vec<VerificationResult> {
+        let mut results = Vec::new();
+        for credential in credentials_to_verify {
+            results.push(Self::verify_credential(credential));
```

Comment:
So, the semantics here are that if any credential fails to verify the whole batch fails? Maybe that's fine, but we might have to rethink it when we start to integrate into real workflows.

Reply:
Actually no, the results are returned as a vec for each verification future. Once the batch_verify completes the caller can sort through the Results for `Ok`'s and `Err`'s. 

[join_all](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.16/futures/future/fn.join_all.html)